astropy/coordinates/sky_coordinate.py

```diff
@@ -26,6 +27,63 @@
 __all__ = ['SkyCoord', 'SkyCoordInfo']
 
 
+@contextlib.contextmanager
+def _set_time_writeable_true(obj):
+    """Context manager to allow __setitem__ to operate on frame attributes.
+
+    Not used in current implementation which just insists that they are
+    equivalent but see commit 4e3f99943.
+    """
+    is_time = isinstance(obj, Time)
+    try:
+        if is_time:
+            writeable = obj.writeable
+            obj.writeable = True
+        yield
+    finally:
+        if is_time:
+            obj.writeable = writeable
+
+
+def _frame_attribute_setitem(frame, attr, item, value):
+    """Set a frame attribute ``frame.attr[item] = value```
+
+    This is not used in the current implementation of __setitem__ but here in
+    the WIP for reference of what might be possible.
+
+    The trick is that frame attributes can be scalars and they might need to be
+    broadcast to the SkyCoord shape.  For example if a SkyCoord ``sc1`` has
+    shape=(3,) with scalar equinox='B1999', and ``sc2`` has equinox='B2015',
+    then setting ``sc1[1] = sc2[2]`` needs to make sc1.equinox into shape=(3,)
+    and result in ['B1999', 'B2015', 'B1999'].
+
+    It isn't obvious if this can cause problems, so don't do that without
+    some discussion.
+    """
+    _attr = '_' + attr
+    frame_attr = getattr(frame, _attr)
+    value_attr = getattr(value, _attr)
+    frame_attr_shape = getattr(frame_attr, 'shape', ())
+
+    # If all the same then do nothing
+    if np.all(frame_attr == value_attr):
+        return
+
+    # At this point we need to ensure frame_attr has the same shape as frame
+    # because ``item`` is referenced to frame shape and we will set frame_attr[item].
+    if frame_attr_shape == ():
+        frame_attr = frame_attr.take(np.zeros(frame.shape))
+        setattr(frame, _attr, frame_attr)
+
+    elif frame_attr_shape != frame.shape:
+        # How to broadcast a ShapedLikeNDArray subclass like Time. E.g.
+        # an obstime attribute with shape (3,) and the frame has shape (10, 3).
+        raise NotImplementedError('help me @mhvk!')
```

Comment:
The attribute `__get__` should have ensured the attribute is either a scalar or has been broadcast to the frame shape: 
https://github.com/astropy/astropy/blob/60fb417789abd03269f4d3af412ee2004983e869/astropy/coordinates/attributes.py#L110-L114
But that of course may leave it unwriteable...  I guess the safest might be to do the same broadcasting (sadly, two different ways to do it until I get around to #8610) and then copy.